### PR TITLE
Elastic Net Regularizer

### LINKED ISSUE #47: L2 regularization scaling is inconsistent
If we use define `l2` regularization as:
```
min_x ( f(x) + lambda / 2 \|x\|_2^2 )
```
then the regularizer is given by
```
1/2 \|x\|_2^2
```
which has proximal operator `1 / (1 +t) x` (which is what we [currently have implemented](https://github.com/dask/dask-glm/blob/master/dask_glm/regularizers.py#L10)).  However, all of our [other methods](https://github.com/dask/dask-glm/blob/master/dask_glm/regularizers.py#L24) assume the regularizer is given by `\|x\|_2^2` (without the 1/2 scaling).  

This doesn't affect quality, but we should be consistent.

cc @postelrich

### LINKED ISSUE #48: Abstract Base Classes
We should probably implement abstract base classes for `Regularizer` and `Family` with corresponding documentation; as #47 and #46 have highlighted, the lack of a solid template can lead to problems.

- [x] Regularizer Base Class
- [ ] Family Base Class

### PR DESCRIPTION
**Pull request by postelrich**
This PR adds elastic net regularization, the weighted sum of l1 and l2, as an option.  Couldn't find a proximal operator so derived it with much much help from @moody-marlin in an included notebook. It also adds an abstract base class for `Regularizer` and fixes the formulas for l2.

Resolves #48 
Resolves #47

### CODE REVIEW

File: dask_glm/algorithms.py [hunk 1]
```diff
@@ -319,7 +319,7 @@ def bfgs(X, y, max_iter=500, tol=1e-14, family=Logistic):
     return beta
 
 
-def proximal_grad(X, y, regularizer=L1, lamduh=0.1, family=Logistic,
+def proximal_grad(X, y, regularizer=L1(), lamduh=0.1, family=Logistic,
```
Comment (TomAugspurger) [inline dask_glm/algorithms.py]: Thoughts about changing this to `'l1`' and looking it up in `_regularizers`?

Comment (postelrich) [inline dask_glm/algorithms.py]: I agree, should be change.  Also in addition, we could move the string logic to a function in `regularizers.py`

**Comment by postelrich** [comment]
@TomAugspurger What do you think of my latest commit?  I use the base to retrieve child classes via string.

File: dask_glm/regularizers.py [hunk 2]
```diff
@@ -9,6 +9,7 @@ class Regularizer(object):
     Defines the set of methods required to create a new regularization object. This includes
     the regularization functions itself and it's gradient, hessian, and proximal operator.
     """
+    _name = '_base'
```
Comment (TomAugspurger) [inline dask_glm/regularizers.py]: I think maybe this should be just `name`. I think @moody-marlin mentioned that users might defining their own regularizers, so they shouldn't need to override private attributes.

File: dask_glm/algorithms.py [hunk 3]
```diff
@@ -150,12 +150,12 @@ def newton(X, y, max_iter=50, tol=1e-8, family=Logistic):
...
 
 
-def admm(X, y, regularizer=L1, lamduh=0.1, rho=1, over_relax=1,
+def admm(X, y, regularizer='l1', lamduh=0.1, rho=1, over_relax=1,
          max_iter=250, abstol=1e-4, reltol=1e-2, family=Logistic):
 
     pointwise_loss = family.pointwise_loss
     pointwise_gradient = family.pointwise_gradient
-    regularizer = _regularizers.get(regularizer, regularizer)  # string
+    regularizer = Regularizer.get(regularizer)
```
Comment (moody-marlin) [inline dask_glm/algorithms.py]: I'll be sad to see this line go, but :+1:

File: dask_glm/regularizers.py [hunk 4]
```diff
@@ -3,85 +3,123 @@
...
 
-class L2(object):
+class Regularizer(object):
+    """Abstract base class for regularization object.
 
-    @staticmethod
-    def proximal_operator(beta, t):
-        return 1 / (1 + t) * beta
+    Defines the set of methods required to create a new regularization object. This includes
+    the regularization functions itself and it's gradient, hessian, and proximal operator.
```
Comment (moody-marlin) [inline dask_glm/regularizers.py]: it's -> its

File: dask_glm/regularizers.py [hunk 5]
```diff
@@ -3,85 +3,123 @@
...
+
+    def hessian(self, beta):
+        """Hessian of regularization function."""
+        raise NotImplementedError
 
-    @staticmethod
-    def f(beta):
-        return (beta**2).sum()
+    def proximal_operator(self, beta, t):
+        """Proximal operator function for non-differentiable regularization function."""
```
Comment (moody-marlin) [inline dask_glm/regularizers.py]: ```Proximal operator for regularization function```; the regularizer doesn't need to be non-differentiable, it just *can* be.

File: dask_glm/regularizers.py [hunk 6]
```diff
@@ -3,85 +3,123 @@
...
         else:
             return np.sign(beta)
 
-    @staticmethod
-    def add_reg_grad(grad, lam):
-        def wrapped(beta, *args):
-            return grad(beta, *args) + lam * L1.gradient(beta)
-        return wrapped
+    def hessian(self, beta):
+        raise ValueError('l1 norm is not twice differentiable!')
```
Comment (moody-marlin) [inline dask_glm/regularizers.py]: We should probably just fix this with this PR; this should be similar to the gradient:
```
if np.any(np.isclose(beta, 0)):
    raise ValueError('l1 norm is not twice differentiable at 0!')
else:
    return np.zeros((beta.shape[0], beta.shape[0]))
```
l1 regularization is a straight line everywhere except at 0 where there's a kink.

File: dask_glm/tests/test_algos_families.py [hunk 7]
```diff
@@ -89,7 +89,7 @@ def test_basic_unreg_descent(func, kwargs, N, nchunks, family):
 @pytest.mark.parametrize('nchunks', [1, 10])
 @pytest.mark.parametrize('family', [Logistic, Normal, Poisson])
 @pytest.mark.parametrize('lam', [0.01, 1.2, 4.05])
-@pytest.mark.parametrize('reg', [L1, L2])
+@pytest.mark.parametrize('reg', [r() for r in Regularizer.__subclasses__()])
```
Comment (moody-marlin) [inline dask_glm/tests/test_algos_families.py]: Nice.

File: dask_glm/regularizers.py (hunk 2 above)
Comment (postelrich) [inline dask_glm/regularizers.py]: I agree, made the change.

File: dask_glm/regularizers.py (hunk 6 above)
Comment (postelrich) [inline dask_glm/regularizers.py]: Then do we want to switch the elastic net hessian to include the l1 side to the weight?  It won't have an effect except raise when there are errors.

Comment (moody-marlin) [inline dask_glm/regularizers.py]: Ah yes, definitely - I missed that.

**Comment by postelrich** [comment]
@moody-marlin ok implemented your changes, looks good to go.

**Comment by moody-marlin** [comment]
@TomAugspurger do you have any other comments / suggestions?  LGTM.

**Comment by TomAugspurger** [comment]
👍  on a quick skim.
